src/base/dev/pic/pic.c

```diff
@@ -584,14 +584,20 @@ void run_irqs(void)
         * irq code actually runs, will reset the bits.  We also reset them here,
         * since dos code won't necessarily run.
         */
-       while((local_pic_ilevel = pic_get_ilevel()) != -1) { /* while something to do*/
+       if((local_pic_ilevel = pic_get_ilevel()) != -1) { /* if something to do*/
                pic_print(1, "Running irq lvl ", local_pic_ilevel, "");
                clear_bit(local_pic_ilevel, &pic_irr);
 	       /* pic_isr bit is set in do_irq() */
                ret = (pic_iinfo[local_pic_ilevel].func ?
 	    	      pic_iinfo[local_pic_ilevel].func(local_pic_ilevel) : 1);      /* run the function */
 	       if (ret) {
 		       do_irq(local_pic_ilevel);
+		       if (pic_pending() && !isset_IF())
```

Comment:
Such check hardly makes any sense:
do_irq() calls real_run_int() so IF is going
to be clear in any case.

Reply:
do_irq indeed usually calls real_run_int or run_pm_int which clears IF but I saw it can call a callback instead (for the ipx/pkt networking code) which doesn't disable ints as for as I can see.

Reply:
We seem to have only 2 call-backs:
pkt_receiver_callback() that does `fake_int_to()`
and ipx_esr_call() that does `do_call_back()` to
complete the call-back in one go.
So it seems like the first one clears IF, and
the second one returns with irq already serviced,
in which case the `while` loop may make sense,
to schedule something else. But I'd rather have
no case 2, and once its not there, apply your patch
(or something similar). But if we get rid of case 2,
the IF will always be cleared.
So what would be the best strategy here, on your
opinion?

Reply:
No-no, wrong.
We do not have the case 2 already.
`ipx_esr_irq_begin()` does `fake_int_to()`.
So I don't think we ever have IF set.

Reply:
Also the comment might help, like
"pic_pending() can happen here only in special mask mode".
Otherwise I am sure people would get quite confused.

Reply:
There is a comment there already, it's just hidden a bit here. However checking myself pkt_receiver_callback does not clear IF if p_helper_size==0. If that check can be eliminated we could just move all the relevant fake_int_to calls to pic.c to make it clearer.

Reply:
Yes, lets just assert on that.
pkt_check_receive() won't allow the irq
unless setting p_helper_size.

Reply:
All done I think.